File: CHUONG_5/5.06_ListView_Advanced/app/src/main/java/com/example/myapplication/data/LocalDataSource.java
package com.example.myapplication.data;

import android.content.res.Resources;

import com.example.myapplication.R;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

// Read country.json, convert JSON to Object Java:
public class LocalDataSource {
    private final Resources mResources;

    public LocalDataSource(Resources resources){
        mResources = resources;
    }

    public List<Country> loadCountries(){
        String jsonData = readJsonFile();
        ObjectMapper mapper = new ObjectMapper();
        try {
            return mapper.readValue(jsonData, new TypeReference<List<Country>>() {});
        } catch (Exception exception) {
            return new ArrayList<>();
        }
    }
    private String readJsonFile(){
        StringWriter writer = new StringWriter();
        char[] buffer = new char[1024];
        try (InputStream inputStream = mResources.openRawResource(R.raw.country)) {
            Reader reader = new BufferedReader(
                    new InputStreamReader(inputStream, StandardCharsets.UTF_8)
            );
            int size;
            while ((size = reader.read(buffer)) != -1){
                writer.write(buffer, 0, size);
            }
        } catch (Exception ex) {

        }
        return writer.toString();
    }
}

File: CHUONG_10/10.2_ex1/app/src/main/java/com/example/myapplication/data/LocalDataSource.kt
package com.example.myapplication.data

import android.content.res.Resources
import com.example.myapplication.R
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.io.BufferedReader
import java.io.InputStreamReader

class LocalDataSource (
    private val resources: Resources
) {
    private val countries: MutableList<Country> = mutableListOf()

    companion object {
        @Volatile private var instance: LocalDataSource? = null

        fun getInstance(resources: Resources): LocalDataSource =
            instance ?: synchronized(this) {
                instance ?: LocalDataSource(resources).also { instance = it }
            }
    }

    fun getCountries(): List<Country> {
        if (countries.isEmpty()) {
            val json = readJsonFile()
            val mapper = jacksonObjectMapper()
            countries.addAll(mapper.readValue(json))
        }
        return countries
    }

    private fun readJsonFile(): String {
        val inputStream = resources.openRawResource(R.raw.country)
        return BufferedReader(InputStreamReader(inputStream)).use {it.readText()}
    }
    fun getCountry(countryName: String): Country? {
        if (countries.isEmpty()) {
            getCountries()
        }
        return countries.find { it.name == countryName }
    }
}